DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Helpers/ParameterValueExtractor.cs

```diff
@@ -0,0 +1,186 @@
+  {
+    var intVal = GetValueGeneric<int?>(
+      element,
+      builtInParameter,
+      StorageType.Integer,
+      (parameter) => parameter.AsInteger()
+    );
+
+    return intVal.HasValue ? Convert.ToBoolean(intVal.Value) : null;
+  }
+
+  public string? GetValueAsString(Element element, BuiltInParameter builtInParameter)
+  {
+    return GetValueGeneric(element, builtInParameter, StorageType.String, (parameter) => parameter.AsString());
+  }
+
+  private string? GetValueAsString(Parameter parameter)
+  {
+    return GetValueGeneric(parameter, StorageType.String, (parameter) => parameter.AsString());
+  }
+
+  public ElementId? GetValueAsElementId(Element element, BuiltInParameter builtInParameter)
+  {
+    return GetValueGeneric(element, builtInParameter, StorageType.ElementId, (parameter) => parameter.AsElementId());
+  }
+
+  public ElementId? GetValueAsElementId(Parameter parameter)
+  {
+    return GetValueGeneric(parameter, StorageType.ElementId, (parameter) => parameter.AsElementId());
+  }
+
+  public T GetValueAsDocumentObject<T>(Element element, BuiltInParameter builtInParameter)
+    where T : class
+  {
+    ElementId? elementId = GetValueAsElementId(element, builtInParameter);
+    Element paramElement = element.Document.GetElement(elementId);
+    return paramElement as T
+      ?? throw new SpeckleConversionException(
+        $"Unable to cast retrieved element of type {paramElement.GetType()} to an element of type {typeof(T)}"
+      );
+  }
+
+  private TResult? GetValueGeneric<TResult>(
+    Element element,
+    BuiltInParameter builtInParameter,
+    StorageType expectedStorageType,
+    Func<DB.Parameter, TResult> getParamValue
+  )
+  {
+    if (!_uniqueIdToUsedParameterSetMap.TryGetValue(element.UniqueId, out HashSet<BuiltInParameter> usedParameters))
+    {
+      usedParameters = new();
+      _uniqueIdToUsedParameterSetMap[element.UniqueId] = usedParameters;
+    }
+    usedParameters.Add(builtInParameter);
+    var parameter = element.get_Parameter(builtInParameter);
+    return GetValueGeneric(parameter, expectedStorageType, getParamValue);
+  }
+
+  private TResult? GetValueGeneric<TResult>(
+    Parameter parameter,
+    StorageType expectedStorageType,
+    Func<DB.Parameter, TResult> getParamValue
+  )
+  {
+    if (parameter == null || !parameter.HasValue)
+    {
+      return default;
+    }
+
+    if (parameter.StorageType != expectedStorageType)
+    {
+      throw new SpeckleConversionException(
+        $"Expected parameter of storage type {expectedStorageType} but got parameter of storage type {parameter.StorageType}"
+      );
+    }
+
+    return getParamValue(parameter);
+  }
+
+  public Dictionary<string, Parameter> GetAllRemainingParams(DB.Element revitElement)
+  {
+    var allParams = new Dictionary<string, Parameter>();
+    AddElementParamsToDict(revitElement, allParams);
+
+    //var elementType = revitElement.Document.GetElement(revitElement.GetTypeId());
```

Comment:
Should this be commented out?

Reply:
In my opinion, yes, but this is how our existing converter does things so you're right. I put the type parameters back.